tools/StaticAnalysis/BreakingChangeAnalyzer/BreakingChangeAnalyzer.cs

```diff
@@ -0,0 +1,968 @@
+                if (!found)
+                {
+                    issueLogger.LogBreakingChangeIssue(
+                        cmdlet: cmdlet,
+                        severity: 0,
+                        problemId: ProblemIds.BreakingChangeProblemId.RemovedValidateSetValue,
+                        description: string.Format(Properties.Resources.RemovedValidateSetValueDescription, 
+                            oldParameter.Name, oldValue, cmdlet.Name),
+                        remediation: string.Format(Properties.Resources.RemovedValidateSetValueRemediation, 
+                            oldValue, oldParameter.Name));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Check if the parameter now supports the ValidateNotNullOrEmpty attribute
+        /// </summary>
+        /// <param name="cmdlet"></param>
+        /// <param name="oldParameter"></param>
+        /// <param name="newParameter"></param>
+        /// <param name="issueLogger"></param>
+        private void CheckForValidateNotNullOrEmpty(
+            CmdletBreakingChangeMetadata cmdlet,
+            ParameterMetadata oldParameter,
+            ParameterMetadata newParameter,
+            ReportLogger<BreakingChangeIssue> issueLogger)
+        {
+            // If the parameter didn't have the ValidateNotNullOrEmpty attribute in the
+            // old assembly, but has it in the new assembly, log an issue
+            if (!oldParameter.ValidateNotNullOrEmpty && newParameter.ValidateNotNullOrEmpty)
+            {
+                issueLogger.LogBreakingChangeIssue(
+                        cmdlet: cmdlet,
+                        severity: 0,
+                        problemId: ProblemIds.BreakingChangeProblemId.AddedValidateNotNullOrEmpty,
+                        description: string.Format(Properties.Resources.AddedValidateNotNullOrEmptyDescription, 
+                            oldParameter.Name, cmdlet.Name),
+                        remediation: string.Format(Properties.Resources.AddedValidateNotNullOrEmptyRemediation, 
+                            oldParameter.Name));
+            }
+        }
+
+        /// <summary>
+        /// Compare two types by recursively checking their properties and property
+        /// types, making sure that nothing has been removed or changed.
+        /// </summary>
+        /// <param name="cmdlet"></param>
+        /// <param name="oldType"></param>
+        /// <param name="newType"></param>
+        /// <param name="issueLogger"></param>
+        private void CompareTypes(
+            CmdletBreakingChangeMetadata cmdlet,
+            TypeMetadata oldType,
+            TypeMetadata newType,
+            ReportLogger<BreakingChangeIssue> issueLogger)
+        {
+            // For each property in the old assembly type, find the corresponding
+            // property in the new assembly type
+            foreach (var oldProperty in oldType.Properties.Keys)
```

Comment:
In `ModuleMetadata`, the type dictionary is initialized with all of the primitive types, so they will never be passed through as arguments to this method